File: typescript/2023/src/day7/day7.ts
import { readAsStringArray } from "../utils/getInput";

type Hand = {
  hand: string;
  bid: number;
  score: number;
};

const rankHands = (input: string[], handleJoker: boolean = false): Hand[] => {
  let hands = input
    .map((line) => line.split(" "))
    .map((hand) => {
      let handMap = new Map();
      let hasJoker = false;
      hand[0].split("").forEach((c) => {
        if (c === "J") hasJoker = true;
        if (handMap.has(c)) handMap.set(c, handMap.get(c) + 1);
        else handMap.set(c, 1);
      });
      if (handleJoker && hasJoker) handMap = replaceJokers(handMap);
      return {
        hand: hand[0],
        bid: parseInt(hand[1]),
        score: scoreHand(handMap),
      };
    });
  hands = hands.sort((a, b) => {
    if (a.score === b.score) {
      return compareCards(a.hand, b.hand, handleJoker);
    } else return a.score - b.score;
  });
  return hands;
};

const replaceJokers = (hand: Map<string, number>): Map<string, number> => {
  let lv = 0;
  let lk = "";
  hand.forEach((v, k) => {
    if (v > lv && k !== "J") {
      lv = v;
      lk = k;
    }
  });
  let jv = hand.get("J");
  if (jv === undefined) throw Error;
  hand.delete("J");
  hand.set(lk, lv + jv);
  return hand;
};

const compareCards = (
  card1: string,
  card2: string,
  handleJoker: boolean = false,
): number => {
  const c1 = Array.from(card1);
  const c2 = Array.from(card2);
  const cardStrength = handleJoker ? "J23456789TQKA" : "23456789JTQKA";

  for (let i = 0; i < c1.length; i++) {
    if (cardStrength.indexOf(c1[i]) > cardStrength.indexOf(c2[i])) return 1;
    else if (cardStrength.indexOf(c1[i]) < cardStrength.indexOf(c2[i]))
      return -1;
  }

  return 0;
};

const scoreHand = (hand: Map<string, number>) => {
  let score = 0;
  if (hand.size === 1) score = 7;
  else if (hand.size === 2) {
    hand.forEach((v, _) => {
      if (v === 4) score = 6;
      else if (v === 3) score = 5;
    });
  } else if (hand.size === 3) {
    hand.forEach((v, _) => {
      if (v === 3) score = 4;
      else if (v === 2) score = 3;
    });
  } else if (hand.size === 4) score = 2;
  else if (hand.size === 5) score = 1;
  return score;
};

export const partOne = (filename: string): number => {
  const input = readAsStringArray(filename);
  const hands = rankHands(input);
  return hands
    .map((hand, index) => {
      return hand.bid * (index + 1);
    })
    .reduce((acc, val) => acc + val);
};

export const partTwo = (filename: string): number => {
  const input = readAsStringArray(filename);
  const hands = rankHands(input, true);
  return hands
    .map((hand, index) => {
      return hand.bid * (index + 1);
    })
    .reduce((acc, val) => acc + val);
};

File: golang/2023/day7/day7.go
package day7

import (
	"aoc/golang/utils"
	"fmt"
	"log"
	"sort"
	"strconv"
	"strings"
)

type Hand struct {
	hand  string
	bid   int
	score int
}

func rankHands(input []string, joker bool) []Hand {
	hands := make([]Hand, len(input))

	for i, round := range input {
		tmp := strings.Fields(round)
		hand := tmp[0]
		bid, err := strconv.Atoi(tmp[1])
		if err != nil {
			log.Fatal("Could not convert string to int", err)
		}

		handMap := make(map[rune]int)
		for _, c := range hand {
			handMap[c] += 1
		}

		if joker {
			replaceJokers(handMap)
		}

		score := scoreHand(handMap)
		hands[i] = Hand{
			hand,
			bid,
			score,
		}
	}

	// The closure evalueates if hands[i] < hands[j]
	sort.Slice(hands, func(i, j int) bool {
		var cardStrengths = "AKQJT98765432"
		if joker {
			cardStrengths = "AKQT98765432J"
		}
		hand1 := hands[i]
		hand2 := hands[j]

		if hand1.score < hand2.score {
			return true
		} else if hand1.score == hand2.score {
			for c := 0; c < len(hand1.hand); c++ {
				cs1 := strings.IndexByte(cardStrengths, hand1.hand[c])
				cs2 := strings.IndexByte(cardStrengths, hand2.hand[c])

				if cs1 > cs2 {
					return true
				} else if cs1 < cs2 {
					return false
				}
			}
		}
		return false
	})

	return hands
}

func scoreHand(hand map[rune]int) (score int) {
	switch len(hand) {
	case 1:
		score = 7
	case 2:
		for _, v := range hand {
			if v == 4 {
				score = 6
			} else if v == 3 {
				score = 5
			}
		}
	case 3:
		for _, v := range hand {
			if v == 3 {
				score = 4
			} else if v == 2 {
				score = 3
			}
		}
	case 4:
		score = 2
	case 5:
		score = 1
	}
	return
}

func replaceJokers(hand map[rune]int) {
	largeKey := rune(0)
	largeVal := 0

	for k, v := range hand {
		if v > largeVal && k != 'J' {
			largeVal = v
			largeKey = k
		}
	}

	jokerVal := hand['J']
	hand[largeKey] += jokerVal
	delete(hand, 'J')
}

func runPartOne(filename string) int {
	fmt.Println("Part One")
	input := utils.ReadAsSlice[string](filename)
	hands := rankHands(input, false)
	score := 0
	for i, h := range hands {
		score += h.bid * (i + 1)
	}
	return score
}

func runPartTwo(filename string) int {
	fmt.Println("Part Two")
	input := utils.ReadAsSlice[string](filename)
	hands := rankHands(input, true)
	score := 0
	for i, h := range hands {
		score += h.bid * (i + 1)
	}
	return score
}
